static-analysis-plugin/plugin/src/test/groovy/com/novoda/test/TestProject.groovy

```diff
@@ -0,0 +1,111 @@
+package com.novoda.test
+
+import org.gradle.testkit.runner.BuildResult
+import org.gradle.testkit.runner.GradleRunner
+
+abstract class TestProject {
+    private static final Closure<String> EXTENSION_TEMPLATE = { TestProject project ->
+        """
+staticAnalysis {
+    ${formatPenalty(project)}
+}
+"""
+    }
+
+    private final File projectDir
+    private final GradleRunner gradleRunner
+    private final Closure<String> template
+    Map<String, List<File>> sourceSets = [main: []]
+    String penalty
+
+    TestProject(Closure<String> template) {
+        this.template = template
+        this.projectDir = createProjectDir("${System.currentTimeMillis()}")
+        this.gradleRunner = GradleRunner.create()
+                .withProjectDir(projectDir)
+                .withDebug(true)
+                .withPluginClasspath()
+                .forwardStdOutput(new OutputStreamWriter(System.out))
+                .forwardStdError(new OutputStreamWriter(System.out))
+    }
+
+    private static File createProjectDir(String path) {
+        File dir = new File(Fixtures.BUILD_DIR, "test-projects/$path")
+        dir.deleteDir()
+        dir.mkdirs()
+        return dir
+    }
+
+    List<String> defaultArguments() {
+        Collections.emptyList()
+    }
+
+    void copyFile(File source, String path) {
+        File file = new File(gradleRunner.projectDir, path)
+        file.parentFile.mkdirs()
+        file.text = source.text
+    }
+
+    public TestProject withSourceSet(String sourceSet, File... srcDirs) {
+        sourceSets[sourceSet] = srcDirs
+        return this
+    }
+
+    public TestProject withPenalty(String penalty) {
+        this.penalty = "penalty $penalty"
+        return this
+    }
+
+    public Result build(String... arguments) {
+        copyFile(Fixtures.CHECKSTYLE_CONFIG, 'config/checkstyle/checkstyle.xml')
+        new File(projectDir, 'build.gradle').text = template.call(this)
+        BuildResult buildResult = withArguments(arguments).build()
+        createResult(buildResult)
+    }
+
+    public Result buildAndFail(String... arguments) {
+        copyFile(Fixtures.CHECKSTYLE_CONFIG, 'config/checkstyle/checkstyle.xml')
+        new File(projectDir, 'build.gradle').text = template.call(this)
```

Comment:
there's some duplication here, but I reckon this code will change quite fast, quite soon.